cli/tools/check.rs

```diff
@@ -249,70 +245,186 @@ impl TypeChecker {
+      ),
+      pending: VecDeque::new(),
     };
     let TscRoots {
       roots: root_names,
       missing_diagnostics,
-      maybe_check_hash,
+      maybe_hasher,
     } = get_tsc_roots(
-      &self.sys,
-      &self.npm_resolver,
-      &self.node_resolver,
+      &mut graph_walker,
       &graph,
-      check_js,
       check_state_hash(&self.npm_resolver),
-      type_check_mode,
-      &ts_config,
+      options.type_check_mode,
     );
 
-    let missing_diagnostics =
-      missing_diagnostics.filter(filter_remote_diagnostics);
+    let mut missing_diagnostics = missing_diagnostics
+      .filter(|d| self.should_include_diagnostic(options.type_check_mode, d));
+    missing_diagnostics.apply_fast_check_source_maps(&graph);
 
-    if root_names.is_empty() && missing_diagnostics.is_empty() {
-      return Ok((graph.into(), Default::default()));
+    // group every root into a folder
+    struct FolderInfo<'a> {
+      tsconfig: &'a TsConfig,
+      roots: Vec<(Url, MediaType)>,
+      maybe_deno_json: Option<&'a Arc<deno_config::deno_json::ConfigFile>>,
+    }
+
+    let mut root_names_by_config = IndexMap::new();
+    for (url, media_type) in root_names {
+      let folder = tsconfig_resolver.folder_for_specifier(&url);
+      let entry = root_names_by_config
+        .entry(folder.dir.dir_url())
+        .or_insert_with(|| FolderInfo {
+          roots: Vec::new(),
+          tsconfig: &folder.tsconfig,
+          maybe_deno_json: folder.dir.maybe_deno_json(),
+        });
+      entry.roots.push((url, media_type));
+    }
+
+    let type_check_cache =
+      TypeCheckCache::new(self.caches.type_checking_cache_db());
+    let mut diagnostics = missing_diagnostics;
+    let logged_check_roots = deno_core::unsync::Flag::lowered();
+    let log_check_roots = || {
+      if logged_check_roots.raise() {
+        for root in &graph.roots {
+          let root_str = root.as_str();
+          log::info!(
+            "{} {}",
+            colors::green("Check"),
+            to_percent_decoded_str(root_str)
+          );
+        }
+      }
+    };
+
+    if !diagnostics.is_empty() {
+      log_check_roots();
+    }
+
+    for (folder_url, folder) in root_names_by_config {
+      let current_diagnostics = self.check_diagnostics_in_folder(
+        folder.roots,
+        folder.tsconfig,
+        folder.maybe_deno_json.map(|d| &**d),
+        // todo(dsherret): avoid clone here when there's only one folder
+        // clone to ensure we don't share the state across folders
+        graph_walker.clone(),
+        &graph,
+        maybe_hasher.clone(),
+        &type_check_cache,
+        &options,
+        &log_check_roots,
+      )?;
+      for diagnostic in current_diagnostics.into_iter() {
+        if let Some(file_name) = &diagnostic.file_name {
+          let folder = tsconfig_resolver.folder_for_specifier_str(file_name);
+          if folder.dir.dir_url() != folder_url {
+            continue; // not the diagnostic for this folder
```

Comment:
Actually, I think we should change this to display diagnostics grouped by config file with compilerOptions set. Otherwise it becomes confusing why a type might be `any` for example.

Reply:
Done.